src/relax/transform/lift_transform_params.cc

```diff
@@ -37,405 +38,443 @@
+    info_.num_runtime_params = num_runtime_params;
+
+    for (size_t i = num_runtime_params; i < func->params.size(); i++) {
+      liftable_vars_.insert(func->params[i]);
+      for (const auto& tir_var : DefinableTIRVarsInStructInfo(GetStructInfo(func->params[i]))) {
+        liftable_vars_.insert(tir_var);
+      }
+    }
+    ExprVisitor::VisitExpr_(func);
   }
 
- private:
   void VisitBindingBlock_(const DataflowBlockNode* block) final {
+    bool cache = is_in_dataflow_block_;
     is_in_dataflow_block_ = true;
     ExprVisitor::VisitBindingBlock_(block);
-    is_in_dataflow_block_ = false;
+    is_in_dataflow_block_ = cache;
+  }
+
+  void VisitBinding(const Binding& binding) override {
+    if (CanLiftBinding(binding)) {
+      info_.computable_at_compile_time.push_back(binding);
+      liftable_vars_.insert(binding->var);
+    } else {
+      info_.required_at_runtime.insert(binding->var);
+      auto bound_value = GetBoundValue(binding);
+      for (const auto& upstream_var : FreeVars(bound_value)) {
+        info_.required_at_runtime.insert(upstream_var);
+      }
+      for (const auto& tir_var : FreeSymbolicVars(bound_value)) {
+        info_.required_at_runtime.insert(tir_var);
+      }
+    }
   }
 
-  void VisitBinding_(const VarBindingNode* binding) final {
-    bool can_lift = true;
+  bool CanLiftBinding(const Binding& binding) const {
+    auto value = GetBoundValue(binding);
 
     // Cond 1. Do not lift bindings outside dataflow blocks.
     if (!is_in_dataflow_block_) {
-      can_lift = false;
+      return false;
     }
 
     // Cond 2. Do not lift regarding the "builtin.stop_lift_params" op.
-    if (const auto* call = binding->value.as<CallNode>()) {
+    if (const auto* call = value.as<CallNode>()) {
       static const Op& stop_lift_params_op = Op::Get("relax.builtin.stop_lift_params");
       if (call->op.same_as(stop_lift_params_op)) {
-        can_lift = false;
+        return false;
       }
     }
 
     // Cond 3. Do not lift when involving Vars that are not liftable.
-    auto producers = FreeVars(binding->value);
-    bool uses_only_liftable_producers = builder_.UsesOnlyLiftableProducers(binding->value);
-    if (!uses_only_liftable_producers) {
-      can_lift = false;
+    for (const auto& var : FreeVars(value)) {
+      if (!liftable_vars_.count(var)) {
+        return false;
+      }
     }
 
     // Cond 4. Do not lift when its struct info contains symbolic variables that do not appear in
     // params.
     for (const auto& var : TIRVarsInStructInfo(GetStructInfo(binding->var))) {
-      if (!param_symbolic_vars_.count(var)) {
-        can_lift = false;
+      if (!liftable_vars_.count(var)) {
+        return false;
       }
     }
 
     // Cond 5. Do not lift declarations of external functions
-    if (binding->value.as<relax::ExternFuncNode>()) {
-      can_lift = false;
+    if (value.as<relax::ExternFuncNode>()) {
```

Comment:
I'm curious when we ever have an ExternFunc as the RHS of a binding.

Reply:
It happened more often prior to https://github.com/apache/tvm/pull/15900, but rarely happens now.

At some point, I'd like to allow a relax function to specify when `tvm::runtime::Registry::Get` is invoked.  It's legal for a `R.ExternFunc` to appear within the `IRModule::functions` table, on the RHS of a binding, or inline for a `relax::CallNode::op`.  Currently, all three of those are resolved when the `relax::VirtualMachine` is instantiated, but they could be used to represent binding at instantiation-time, at function start-time, and at function execution-time.  In that case, the ExternFunc would become more common as the RHS of a binding.

Reply:
I guess ExternFuncs are permitted to be used in that way per the spec as first-class values, so this is good to handle.